File: sdks/typescript/modulacms-admin-sdk/src/types/deploy.ts
/**
 * Deploy sync types for cross-instance content synchronization.
 *
 * @module types/deploy
 */

/** Health check response from GET /api/v1/deploy/health. */
export type DeployHealthResponse = {
  /** Server status (e.g. "ok"). */
  status: string
  /** ModulaCMS version string. */
  version: string
  /** Unique node identifier for this instance. */
  node_id: string
}

/** Optional request body for POST /api/v1/deploy/export. */
export type DeployExportRequest = {
  /** Table names to export. Omit for default table set. */
  tables?: string[]
}

/** The full sync payload returned by export and consumed by import. */
export type DeploySyncPayload = Record<string, unknown>

/** Per-table error during a sync operation. */
export type DeploySyncError = {
  /** Table where the error occurred. */
  table: string
  /** Phase of the sync: "export", "validate", "truncate", "insert", or "verify". */
  phase: string
  /** Human-readable error message. */
  message: string
  /** Row ID that caused the error, if applicable. */
  row_id?: string
}

/** Result of a sync import operation. */
export type DeploySyncResult = {
  /** Whether the import completed successfully. */
  success: boolean
  /** Whether this was a dry-run (no database changes). */
  dry_run: boolean
  /** Merge strategy used (e.g. "overwrite"). */
  strategy: string
  /** Table names that were affected. */
  tables_affected: string[]
  /** Per-table row counts. */
  row_counts: Record<string, number>
  /** Path to the backup created before import, if any. */
  backup_path: string
  /** Snapshot identifier. */
  snapshot_id: string
  /** Human-readable duration string. */
  duration: string
  /** Errors encountered during the import. */
  errors?: DeploySyncError[]
  /** Warnings generated during the import. */
  warnings?: string[]
}

File: sdks/go/deploy.go
package modula

import (
	"context"
	"encoding/json"
)

// DeployHealthResponse is returned by GET /api/v1/deploy/health.
type DeployHealthResponse struct {
	Status  string `json:"status"`
	Version string `json:"version"`
	NodeID  string `json:"node_id"`
}

// DeployExportRequest is the optional body for POST /api/v1/deploy/export.
type DeployExportRequest struct {
	Tables []string `json:"tables,omitempty"`
}

// DeploySyncResult is returned by POST /api/v1/deploy/import.
type DeploySyncResult struct {
	Success        bool              `json:"success"`
	DryRun         bool              `json:"dry_run"`
	Strategy       string            `json:"strategy"`
	TablesAffected []string          `json:"tables_affected"`
	RowCounts      map[string]int    `json:"row_counts"`
	BackupPath     string            `json:"backup_path"`
	SnapshotID     string            `json:"snapshot_id"`
	Duration       string            `json:"duration"`
	Errors         []DeploySyncError `json:"errors,omitempty"`
	Warnings       []string          `json:"warnings,omitempty"`
}

// DeploySyncError describes a specific failure during a sync operation.
type DeploySyncError struct {
	Table   string `json:"table"`
	Phase   string `json:"phase"`
	Message string `json:"message"`
	RowID   string `json:"row_id,omitempty"`
}

// DeployResource provides deploy sync operations.
type DeployResource struct {
	http *httpClient
}

// Health returns the deploy health status of the server.
func (d *DeployResource) Health(ctx context.Context) (*DeployHealthResponse, error) {
	var result DeployHealthResponse
	if err := d.http.get(ctx, "/api/v1/deploy/health", nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Export exports a sync payload from the server. If tables is non-empty, only those
// tables are included; otherwise the full database is exported.
// The returned json.RawMessage is the opaque SyncPayload suitable for Import.
func (d *DeployResource) Export(ctx context.Context, tables []string) (json.RawMessage, error) {
	var body any
	if len(tables) > 0 {
		body = DeployExportRequest{Tables: tables}
	}
	var result json.RawMessage
	if err := d.http.post(ctx, "/api/v1/deploy/export", body, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// Import imports a sync payload into the server.
// The payload should be the json.RawMessage returned by Export.
func (d *DeployResource) Import(ctx context.Context, payload json.RawMessage) (*DeploySyncResult, error) {
	var result DeploySyncResult
	if err := d.http.post(ctx, "/api/v1/deploy/import", payload, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// DryRunImport performs a dry-run import, returning what would change without writing.
// The payload should be the json.RawMessage returned by Export.
func (d *DeployResource) DryRunImport(ctx context.Context, payload json.RawMessage) (*DeploySyncResult, error) {
	var result DeploySyncResult
	if err := d.http.post(ctx, "/api/v1/deploy/import?dry_run=true", payload, &result); err != nil {
		return nil, err
	}
	return &result, nil
}
